### DISCUSSION TITLE
The best way to modify buffer outside of snippet

### DISCUSSION THREAD

**Discussion by dorage** [L3MON4D3/LuaSnip#1197] [Q&A] [answered]
Hello, I wrote an auto-import code for missing imports.
It adds missing import statements to the first line of the buffer using `vim.api.nvim_buf_get_lines`.
When I use it inside a function node with `vim.schedule`, it inserts the snippet text.
However, the cursor stays in the position it was before inserting the import statements.
What is the best way to manipulate the buffer outside of the snippet while ensuring proper cursor positioning?

``` lua
s(
    {name = "React: useState", trig = "rehs"},
    {
        d(
            0,
            function()
                vim.schedule(
                    function()
                        ls_auto_import.import(
                            {
                                {source = "react", default_modules = {}, modules = {"useState"}}
                            }
                        )
                    end
                )
                return sn(
                    1,
                    fmt(
                        [[
const [<>, use<>] = useState<<<>>>(<>);
	]],
                        {
                            i(1),
                            f(ls_utils.capitalize_first_char, {1}),
                            i(2),
                            i(3)
                        },
                        fmtopt
                    )
                )
            end
        )
    }
)
```

**Comment by L3MON4D3** [comment, marked as answer]
Hi :)
You'd have to store the cursor-position using an extmark, insert text (which will shift the extmark with the text) and then query the new position of the extmark, and set the cursor on top of it (and then delete the extmark to avoid clutter ;) )

This is a bit cumbersome, and I think we should shift some stuff around s.t. lines can be added/removed in the `resolveExpandParams`-callback as long as they're before/after the line with the trigger.

I'll update here once I get to that :)

**Reply by dorage** [reply]
Thanksss
I just restored cursor by `vim.api.nvim_win_set_cursor()` with range of treesitter node and `vim.api.nvim_win_get_cursor`.
it is working well now!

**Comment by dorage** [comment]
I made it to a plugin.
Plugin includes luasnip callback

https://github.com/dorage/ts-manual-import.nvim
